ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContract.java

```diff
@@ -93,9 +96,11 @@ public BytesValue compute(final BytesValue input, final MessageFrame messageFram
     final ReceiveResponse receiveResponse;
     try {
       receiveResponse = enclave.receive(receiveRequest);
-    } catch (final Exception e) {
+    } catch (final EnclaveException e) {
       LOG.debug("Enclave probably does not have private transaction payload with key {}", key, e);
       return BytesValue.EMPTY;
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException("Failed to Process Private Transaction", e);
```

Comment:
This won't work well when mining blocks as we'll be stuck in a loop of `| ERROR | BlockMiner | Block mining threw an unhandled exception. Failed to Process Private Transaction`.

Maybe we should consider catching this at the `MainnetMessageCallProcessor` level and setting the frame state to an exceptional halt?

Reply:
Will an exceptional halt exit the application? If so the user will need to act to bring back up the node. Perhaps we could do something like 5 retries then exit?

Reply:
Sorry! I think I wasn’t clear. The exceptional halt is an EVM state. It will flag that the frame execution finished unsuccessfully and something went wrong. Therefore, the transaction associated with this message is invalid (failed processing).

I’ll dig the details later! We can discuss it in our sync!

Reply:
Exceptional halt doesn't mean the transaction is invalid - it means the transaction is rolled back. It is still valid and can be included in a block.  e.g. out of gas is an exceptional halt condition. EVM processing stops, any storage changes made by the transaction are rolled back and all gas is consumed then processing moves to the next transaction in the block.

Reply:
Ah my bad! We need is to find what is the best way to fail the tx processing w/o discarding it...

Reply:
With this PR it will keep spinning forever with exception being dumped on the console. When the enclave is back up it will process the block again, followed by the PMT and all will be right in the world! With the unhandled exception (as best as I can tell) no changes are committed.

Reply:
So - given the miner may not have an orion running (I assume that's possible?) - should a private marker always pass at mining time?

Reply:
I'm not sure if we can make it pass at mining time. As part of mining a tx we include it in a block and process the tx, right?

I believe the following scenarios are possible:
1. Miner isn't a Besu node (or doesn't have privacy enabled)
2. Miner is a Besu node with privacy enabled (not a participant of the private tx)
3. Miner is a Besu node with privacy enabled (participant of the private tx)

I believe, for scenario 1, it will behave as if you were sending data to an empty address ('cause it won't use the precompile). So it won't try to talk to Orion anyway.

For scenario 2, if the node isn't a participant, it will just skip the transaction (no payload returned).

For scenario 3, it will fetch the payload from Orion. 

In scenario 2 and 3, Orion might be unavailable, therefore, we want things "to fail".

If I'm mining and I can't talk to Orion, I shouldn't add the PMT to the block (I think it should behave as if the transaction is invalid, but not discarding it - similar to future nonce txs - the transaction will be "valid" as soon as Orion is back online).

If I'm importing blocks and I can't talk to Orion when processing a PMT, I should stop syncing and wait for Orion to be back online.

Does that capture the intended behaviour @antonydenyer ?

Reply:
> So - given the miner may not have an orion running (I assume that's possible?) - should a private marker always pass at mining time?

Correct, if you do not have privacy enabled processing should always be successful (normal rules apply with gas etc).